code/modules/mob/living/bot/bot.dm

```diff
@@ -435,11 +438,63 @@
 		if(D.dir == dir)		return 1
 
 	for(var/obj/machinery/door/D in loc)
+		if(loc.pathweight == 1)
+			loc.pathweight = 2	//Slight bias in pathfinding away from airlocks
 		if(!D.density)			continue
 		if(istype(D, /obj/machinery/door/window))
 			if( dir & D.dir )	return !D.check_access(ID)
 
 			//if((dir & SOUTH) && (D.dir & (EAST|WEST)))		return !D.check_access(ID)
 			//if((dir & EAST ) && (D.dir & (NORTH|SOUTH)))	return !D.check_access(ID)
-		else return !D.check_access(ID)	// it's a real, air blocking door
+		else 
+			if(istype(D, /obj/machinery/door/airlock/lift))	//Stop bots committing suicide down the lift shaft, as amusing as it is
+				return !D.check_access(ID) || D.density
+			var/obj/machinery/door/airlock/A = D
+			if(istype(A) && (A.locked || A.isAllPowerLoss() || A.welded))	//Avoid airlocks that are bolted, welded, or have no power
+				return 1
+			return (!D.check_access(ID))	// it's a real, air blocking door
+
+	for(var/obj/structure/railing/D in loc)
+		if(!D.density)			continue
+		if(dir == D.dir)		return 1
 	return 0
+
+
+//Multi-Z AStar Procs
+
+/turf/proc/Euclidean3dDistance(turf/t)
+	var/euclid_dist = sqrt(Square(src.x - t.x)+Square(src.y - t.y)+Square(src.z - t.z))
+	euclid_dist *= (pathweight+t.pathweight)/2
+	return euclid_dist
```

Comment:
More importantly, I just noticed we're only returning _cardinal_ turfs with the adjacent proc anyway, which defeats the whole point of using Euclid... Whoops.

Testing how the bot handles passing all dirs to adjacent with Euclid. If it's wonky, I'll change it to Manhattan + cardinal.

Reply:
It shouldn't be _that_ big of a deal to use Euclid if the adjacents are cardinal - for cardinally adjacent turfs, Euclid == Manhattan. 

The difference only pops up for diagonals (`Euclid((1,1), (0,0)) == sqrt(2)` while `Manhattan((1,1), (0,0)) == 2)` - so Manhattan is biased towards straight lines where possible.

Reply:
Also, in case it comes up later - spessmobs that have full diagonal motion in 3D (e.g. could work for spesscarp) are most accurately measured in yet another way: `max(abs(x2-x1), abs(y2-y1), abs(z2-z1))` (Chebyshev, or chessboard, distance; works for 2D diagonals if you drop the Zs). 

In practice, though, this doesn't matter for bots unless you specifically want them to home in on a target with ruthless, mechanical precision using the absolute shortest path possible - good for e.g. precision-guided missiles, bad for animal mobs.